[WIP] [API] [Slider] Make Slider colors more customizable

romoore opened the pull request:
The Slider should have more targeted properties for performing customization.
Specifically, it should permit:

* `activeTrackColor` - the "filled" portion of the track. Enabled/Disabled
* `inactiveTrackColor` - the "unfilled" portion of the track. Enabled/Disabled.
* `activeTickMarkColor` - the color of tick marks in the "filled" portion of
   the rail/track. Enabled/Disabled.
* `inactiveTickMarkColor` - the color of tick marks in the "unfilled" portion
   of the rail/track. Enabled/Disabled.
* `thumbColor` - the color of the thumb when not tapped/pressed.
   Normal/Highlighted/Disabled
* `valueLabelTextColor` - the color of the pop-up value label's text
* `valueLabelBackgroundColor` - the color of the pop-up value label's background

# Important
This is an API review. Please do not merge (it's marked WIP for this reason), but provide feedback on the API.

Closes #3137

Pivotal story: https://www.pivotaltracker.com/story/show/155525171

jverkoey commented:
Can you confirm that https://www.pivotaltracker.com/story/show/155525171 is the story for this PR?

romoore commented:
@jverkoey Yep, that looks correct.

jmdetloff commented on components/Slider/src/MDCSlider.h:
```diff
@@ -49,21 +49,140 @@ IB_DESIGNABLE
  Default color is blue.
  */
-@property(nonatomic, strong, null_resettable) UIColor *color UI_APPEARANCE_SELECTOR;
+@property(nonatomic, strong, null_resettable) UIColor *color NS_UNAVAILABLE;
```
I'm not familiar with using NS_UNAVAILABLE on a property like this. Whats the advantage over just moving this to the .m file?

romoore replied:
This is a pretty old PR, and I should follow the deprecations schedule instead. You're right that this normally is only used for superclass APIs, and in this case we should just delete it long-term.

jmdetloff commented on components/Slider/src/MDCSlider.h:
```diff
@@ -49,21 +49,140 @@ IB_DESIGNABLE
+
+ In general, if a property is not specified for a state, the default is to use the
+ @c UIControlStateNormal value. If the @c UIControlStateNormal value is not set, then the property
+ defaults to a system value. Therefore, at a minimum, you should set the value for the normal state.
```
System value? Would this be better described as a material default?

romoore replied:
I took it from UIButton's docs. I'm happy to change to "...defaults to a material default value."

romoore replied:
Ah, we can't actually use Material colors as defaults, since that would change the default behavior. :-\

jmdetloff reviewed (approved):
Api looks good to me, just a couple questions

Commit:
[Slider] Increase color customization API

MDCSlider should allow customization of the following properties:

* activeTrackColor - the "filled" portion of the rail/track
* activeTrackDisabledColor - the "filled" portion of the rail/track when disabled
* inactiveTrackColor - the "unfilled" portion of the rail/track
* inactiveTrackDisabledColor - the "unfilled" portion of the rail/track when disabled
* activeTickMarkColor - the color of tick marks in the "filled" portion of the rail/track
* tickMarkDisabledColor - the color of tick marks in the "filled" portion of the rail/track when disabled
* inactiveTickMarkColor - the color of tick marks in the "unfilled" portion of the rail/track
i* nactiveTickMarkDisabledColor - the color of tick marks in the "unfilled" portion of the rail/track when disabled
* restingThumbColor - the color of the thumb when not tapped/pressed
* raisedThumbColor - the color of the thumb when tapped/pressed
* disabledThumbColor - the color of the thumb when tapped/pressed
* valueLabelTextColor - the color of the pop-up value label's text
* valueLabelBackgroundColor - the color of the pop-up value label's background

The names/implementation is flexible, as long as each of these colors can be independently modified.

Commit:
Working on tests.

jverkoey commented on components/Slider/src/MDCSlider.h:
```diff
@@ -45,25 +45,128 @@ IB_DESIGNABLE
+ In general, if a property is not specified for a state, the default is to use the
+ @c UIControlStateNormal value. If the @c UIControlStateNormal value is not set, then the property
+ defaults to a default value. Therefore, at a minimum, you should set the value for the
+ normal state.
```
If this is now a property, this comment may be out of date.

jverkoey replied:
Ignore me! I misread the diff.

jverkoey commented on components/Slider/src/MDCSlider.h:
```diff
@@ -45,25 +45,128 @@ IB_DESIGNABLE
+          specific state, this method returns the color associated with the @c UIControlStateNormal
+          state.
+ */
+//- (nullable UIColor *)backgroundTrackTickColor:(UIControlState)state;
```
Will these commented-out APIs be marked as deprecated?

romoore replied:
No, this is where I parked the API review. I'm going to be pushing up the actual changes in a series of PRs.

jverkoey reviewed (changes requested):
Requesting clarification on the commented-out APIs.

romoore commented:
All of the work has been done in subsequent PRs.
